### ISSUE TITLE
Create a way for modules to run async code at startup

### ISSUE THREAD

**Issue by babolivier** [matrix-org/synapse#12433]
Some modules might want to run some async code at startup, to e.g. set up their database tables before any code accessing this table can be called (to avoid races). Currently the only way to do so is to run that code inside `run_in_background`, which can be prone to races.

It looks to me like it would be simple enough to add an async `setup` callback somewhere around [here](https://github.com/matrix-org/synapse/blob/e24ff8ebe3d4119d377355402245947f7de61c00/synapse/app/_base.py#L434-L444) that allows modules to run async code before e.g. listeners start.

**Comment by babolivier** [comment]
> (closed by accident?)

Yes, I don't know what's up with GitHub closing issues and PRs when I hit Ctrl + Enter on a comment.

**Comment by richvdh** [comment]
#13409 discusses some potential pitfalls here (though the TL;DR is that I don't think they are real problems)
